Add the user id to the get create and delete keypair

randomswdev opened the pull request:
For #2145

The capability to create of get the keypair of another user is documented in the Nova Rest API reference:
https://docs.openstack.org/api-ref/compute/?expanded=list-keypairs-detail,show-keypair-details-detail,create-or-import-keypair-detail#create-or-import-keypair

This pull request adds the capability to create or get the keypairs also for another user.

* https://github.com/openstack/nova/blob/master/nova/api/openstack/compute/schemas/keypairs.py#L64-L84
* https://github.com/openstack/nova/blob/master/nova/api/openstack/compute/keypairs.py#L148-L156
* https://github.com/openstack/nova/blob/master/nova/api/openstack/compute/keypairs.py#L175-L182

randomswdev commented:
I think the build failure is not related to my ciode, but is more of an environmental issue. Can someone help me fixing it?

jtopjian commented:
@randomswdev Thanks for submitting this. I encourage you to read our [contributor tutorial](https://github.com/gophercloud/gophercloud/tree/master/docs/contributor-tutorial), specifically Step 3, to gather all of the information needed for this PR.

The way this is being implemented is a pretty large breaking change. Normally I don't mind merging patches that cause breaking changes, key pairs are a core resource of openstack and this PR would require a lot of downstream code to be rewritten. Additionally, since this feature was added in a microversion, it's technically not part of the base API calls.

Adding `UserID` to `CreateOpts` is fine to do, as long as the guidelines [here](https://github.com/gophercloud/gophercloud/blob/master/docs/MICROVERSIONS.md#new-request-fields) are followed.

For the `Get` and `Delete` functions, I suggest creating new functions for `Get` and `Delete` that specifically pass the `userID` field. Perhaps `GetWithUserID` and `DeleteWithUserID`.

Please let me know if you have any questions.

randomswdev commented:
@jtopjian thanks for the comments.

> Adding `UserID` to `CreateOpts` is fine to do, as long as the guidelines [here](https://github.com/gophercloud/gophercloud/blob/master/docs/MICROVERSIONS.md#new-request-fields) are followed.

The UserID field is flagged as omitempty, so should be ok based on the guidelines.

> For the `Get` and `Delete` functions, I suggest creating new functions for `Get` and `Delete` that specifically pass the `userID` field. Perhaps `GetWithUserID` and `DeleteWithUserID`.

I added the `GetWithUserID` and `DeleteWithUserID` functions, and reverted the changes to the interface of `Get` and `Delete`, that are now backward compatible.

Let me know if other changes are needed. I'm monuitoring the openlab/check status: if it fails again, is there anything I can do to fix it?

jtopjian commented:
@randomswdev Thanks!

The OpenLab error might be temporary. If it fails again, you can retrigger it by adding a comment with the single word "recheck".

jtopjian commented:
recheck

jtopjian commented:
@randomswdev The OpenLab error is ignorable - you aren't modifying any of the existing acceptance tests, so it's a problem with OpenLab itself.

Let me know if you're done working on this and if it's ready to be merged.

randomswdev commented:
> Let me know if you're done working on this and if it's ready to be merged.

@jtopjian sorry for the delay in answering. Yes, I'm done with hte implementation, so if it is ok for you I think it can be merged.

randomswdev commented:
recheck

randomswdev commented:
Just rebased on the latest master.

jtopjian commented on openstack/compute/v2/extensions/keypairs/requests.go:
```diff
@@ -76,14 +80,28 @@ func Create(client *gophercloud.ServiceClient, opts CreateOptsBuilder) (r Create
 // Get returns public data about a previously uploaded KeyPair.
 func Get(client *gophercloud.ServiceClient, name string) (r GetResult) {
-	resp, err := client.Get(getURL(client, name), &r.Body, nil)
+	r = GetWithUserID(client, name, "")
```
`Get` should not be coupled to the new `GetWithUserID`. They need to be two separate functions that can work unrelated.

jtopjian commented on openstack/compute/v2/extensions/keypairs/requests.go:
```diff
@@ -76,14 +80,28 @@ func Create(client *gophercloud.ServiceClient, opts CreateOptsBuilder) (r Create
 // Delete requests the deletion of a previous stored KeyPair from the server.
 func Delete(client *gophercloud.ServiceClient, name string) (r DeleteResult) {
-	resp, err := client.Delete(deleteURL(client, name), nil)
+	r = DeleteWithUserID(client, name, "")
```
Similar here, `Delete` should not call `DeleteWithUserID`.

jtopjian commented on openstack/compute/v2/extensions/keypairs/requests.go:
```diff
@@ -76,14 +80,28 @@ func Create(client *gophercloud.ServiceClient, opts CreateOptsBuilder) (r Create
+
+// GetWithUserID returns public data about a previously uploaded KeyPair, owned by a specific userID.
+// If the userID is an empty string, it searches the KeyPairs of the current user.
+func GetWithUserID(client *gophercloud.ServiceClient, name string, userID string) (r GetResult) {
```
Instead of passing in `userID` in the function parameters, create a `GetOpts` struct that contains a `UserID` field. See here for an example: https://github.com/gophercloud/gophercloud/blob/master/openstack/compute/v2/extensions/limits/requests.go

The reason this is done is because the user ID is a query parameter of the URL and not part of the URL itself.

jtopjian commented on openstack/compute/v2/extensions/keypairs/requests.go:
```diff
@@ -76,14 +80,28 @@ func Create(client *gophercloud.ServiceClient, opts CreateOptsBuilder) (r Create
+
+// DeleteWithUserID requests the deletion of a previous stored KeyPair from the server, owned by a specific userID.
+// If the userID is an empty string, it deletes the KeyPair of the calling user.
+func DeleteWithUserID(client *gophercloud.ServiceClient, name string, userID string) (r DeleteResult) {
```
Similar here: there needs to be a `DeleteOpts` struct created: https://github.com/gophercloud/gophercloud/blob/master/openstack/blockstorage/v3/qos/requests.go#L67-L100

jtopjian commented on openstack/compute/v2/extensions/keypairs/urls.go:
```diff
@@ -16,10 +16,14 @@ func createURL(c *gophercloud.ServiceClient) string {
 
-func getURL(c *gophercloud.ServiceClient, name string) string {
-	return c.ServiceURL(resourcePath, name)
+func getURL(c *gophercloud.ServiceClient, name string, userID string) string {
```
By using a `GetOpts` struct, you should be able to revert `getURL` back to the original code and just append the query params.

jtopjian commented on openstack/compute/v2/extensions/keypairs/urls.go:
```diff
@@ -16,10 +16,14 @@ func createURL(c *gophercloud.ServiceClient) string {
 
-func deleteURL(c *gophercloud.ServiceClient, name string) string {
-	return getURL(c, name)
+func deleteURL(c *gophercloud.ServiceClient, name string, userID string) string {
```
Similar here: by using `DeleteOpts`, you should be able to revert this.

jtopjian reviewed (changes requested):
@randomswdev Thanks again for working on this. I've left some comments - please let me know if you have any questions or need any help.

jtopjian commented:
@randomswdev Thank you for your work on this. This work was amended to #2186 and merged.
